task_processing/plugins/mesos/execution_framework.py

```diff
@@ -542,12 +545,30 @@ def statusUpdate(self, driver, update):
         if task_id not in self.task_metadata:
             # We assume that a terminal status update has been
             # received for this task already.
-            log.info('Ignoring this status update because a terminal status \
-                update has been receiced for this task already.')
+            log.info('Ignoring this status update because a terminal status'
+                     'update has been receiced for this task already.')
             driver.acknowledgeStatusUpdate(update)
             return
+
         md = self.task_metadata[task_id]
 
+        # If we attempt to accept an offer that has been invalidated by master
+        # for some reason such as offer has been rescinded or we have exceeded
+        # offer_timeout, then we will get TASK_LOST status update back from
+        # mesos master.
+        if task_state == 'TASK_LOST' and \
+                'REASON_INVALID_OFFERS' == str(update.reason):
+            # This task has not been launched. Therefore, we are going to
+            # reenqueue it. We are not propogating any event up to the
+            # application.
+            log.warning('Received TASK_LOST from mesos master because we'
+                        'attemted to accept an invalid offer.')
+            self.task_metadata = self.task_metadata.discard(task_id)
+            self.enqueue_task(md.task_config)
```

Comment:
i agree that it's infrastructure details and we shouldn't concern users with these issues. what i'd like to avoid is large classes with many responsibilities. execution framework is already 600 lines long and we still have more mesos-specific things to add (reconciliations, etc).

my thinking is:

- executors are very isolated pieces of logic, it's easy to merge them into the framework if we later realize that some layer is an overkill
- on the other hand if some logic is embedded into the framework - over time it will be increasingly complex to rip it out and present in form of executor, execution framework will just snowball out of control

i'd rather do the easy thing early than potentially deal with hard thing later.